sklearn/utils/testing.py

```diff
@@ -372,14 +375,30 @@ def __exit__(self, *exc_info):
     assert_greater = _assert_greater
 
 
+def _assert_sparse_allclose(actual, desired, rtol=1e-7, atol=0, err_msg='',
```

Comment:
I would suggest implementing this function in a similar way to how [numpy's `assert_allclose`](https://github.com/numpy/numpy/blob/v1.9.1/numpy/testing/utils.py#L1247) is implemented for dense arrays. Essentially just replace the inner `compare` function with one suitable for sparse matrices. (If you can leverage `np.allclose` it would be ideal. Note that doing `A-B` might raise `EfficiencyWarning`s.)

Also, why not have a helper `assert_safe_sparse_allclose` (or `safe_sparse_assert_allclose`) which works both for dense and sparse data, in the spirit of `safe_sparse_dot` etc.

But at the end of the day, since these asserts are always only used in the test suite, where the matrices that occur are not large, it might just be better (simpler, safer) to cast the arrays to dense. How are sparse matrices being compared in other test cases in the codebase currently?